### PR TITLE
Delay the error on recovery for steady-state executioners for better modular design

### LINKED ISSUE #22432: Delay the recovery early return for steady-state executioners
## Reason
Currently `Eigenvalue::init` and `Steady::init` issue a message and do an early return when application is in recovery mode `if (_app.isRecovering())`. This prevents me from deriving an executioner for doing sequential steady-state calculations with recovery.

## Design
I think recovery mode is allowed even we are at steady-state. I also think we can possibly get rid of the check. But I do not want to go that far and just propose to delay the check in their `execute` function.

## Impact
Better modular design in executioners for allowing C++ inheritance.

### PR DESCRIPTION
**Pull request by YaqiWang**
Close #22432.

### CODE REVIEW

File: test/tests/transfers/multiapp_transfer_transformation/tests
Comment (YaqiWang) [inline test/tests/transfers/multiapp_transfer_transformation/tests:13]: Although this test is using Steady executioner and should be run in recover test, it should not fail. Anyone has insight on this? This is happening when using Steady executioner and try to do recover.

Comment (YaqiWang) [reply test/tests/transfers/multiapp_transfer_transformation/tests:13]: I think I know the reason. When we are doing recovery, the multiapp mesh is not uniformly refined even `Mesh/uniform_refine` is greater than zero. Looks like sub-app mesh is not stored in the check point file. I will add an error check on this. I also attempted a fix:

```
diff --git a/framework/src/actions/SetupMeshCompleteAction.C b/framework/src/actions/SetupMeshCompleteAction.C
index 8f27bb4650..dfc60cedcc 100644
--- a/framework/src/actions/SetupMeshCompleteAction.C
+++ b/framework/src/actions/SetupMeshCompleteAction.C
@@ -63,8 +63,10 @@ SetupMeshCompleteAction::act()
      * file based restart and we need uniform refinements, we'll have to postpone
      * those refinements until after the solution has been read in.
      */
-    if (_app.getExodusFileRestart() == false && _app.isRecovering() == false)
+    if (_app.getExodusFileRestart() == false)
     {
+      if (_app.isRecovering() == false || !_app.isUltimateMaster())
+      {
       TIME_SECTION("uniformRefine", 2, "Uniformly Refining");
 
       if (_mesh->uniformRefineLevel())
@@ -86,6 +88,7 @@ SetupMeshCompleteAction::act()
           _displaced_mesh->update();
         }
       }
+      }
     }
   }
   else if (_current_task == "delete_remote_elements_after_late_geometric_ghosting")
```
which does fixed this problem. But I think it is safe to rather issue the error for now.

**Comment by YaqiWang** [comment]
This turns out more difficult than I thought. Is there a way to invoke app recovery tests? I suppose this change will possibly fail applications.

**Comment by GiudGiud** [comment]
To invoke those tests locally you pass —recover on the command line

**Comment by YaqiWang** [comment]
> To invoke those tests locally you pass —recover on the command line after ./run_tests

Yeah, but I want to test this for all applications.

**Comment by YaqiWang** [comment]
I think this is ready for review. Another easier solution for my need is to add a virtual function like `bool allowRecovery()` in `Eigenvalue` which returns false by default but allows derived class to override. But I think the changes here are preferable. If you do not like the changes in the source of the two steady-state executioners, I can do the virtual function. The changes in tests and the additional error message are probably good.

File: modules/fluid_properties/test/tests/two_phase_fluid_properties_independent/tests
Comment (lindsayad) [inline modules/fluid_properties/test/tests/two_phase_fluid_properties_independent/tests:26]: It didn't used to though

**Review by lindsayad — CHANGES_REQUESTED** [review]

File: modules/fluid_properties/test/tests/two_phase_fluid_properties_independent/tests
Comment (YaqiWang) [reply modules/fluid_properties/test/tests/two_phase_fluid_properties_independent/tests:26]: It did not but this test has Steady executioner, thus recover does not make sense to it based on the current code.

**Comment by YaqiWang** [comment]
I am pretty sure it is the other way around: we are not saving the subapp mesh in the checkpoint file and we are not doing the uniform refinement.

If we still run this test in recover, we will hit the error message I added. I think the other few place to disable recover are legitimate because we are not thoughtful. I did not check but I think there should be places in the repo where steady-state executioners are used and we still run recover test.

**Comment by lindsayad** [comment]
It's nice to comment in the thread

> we are not saving the subapp mesh in the checkpoint file and we are not doing the uniform refinement.

I think you are right. What was failing specifically?

**Comment by YaqiWang** [comment]
> It's nice to comment in the thread
> 
> > we are not saving the subapp mesh in the checkpoint file and we are not doing the uniform refinement.
> 
> I think you are right. What was failing specifically?

Right. Github posted your comments twice, one in the thread another at the end of the other thread which I cannot reply. I only noticed the later or maybe the former was not posted on the page when I was checking

The failing is a binary loading error in `SystemBase::dataLoad`, saying a solution name is not found with non-ascii letters populating the screen because the solution vector name is loaded wrongly due to different solution vector size with different mesh.

File: modules/fluid_properties/test/tests/two_phase_fluid_properties_independent/tests
Comment (lindsayad) [reply modules/fluid_properties/test/tests/two_phase_fluid_properties_independent/tests:26]: Haha I don't think it's fair to say this test does not make sense to run in recover because it's `Steady` while in this PR you are deliberately modifying the `Steady` class so that you can do two sequential steady calculations during recover (at least that is my understanding of your motivation for this PR based on the issue you're referencing)

**Comment by lindsayad** [comment]
> The failing is a binary loading error in `SystemBase::dataLoad`, saying a solution name is not found with non-ascii letters populating the screen because the solution vector name is loaded wrongly due to different solution vector size with different mesh. After I added the error, the failure will be captured by that error.

I see. And we did not hit this before because we never proceeded to read the restartable data in `FEProblemBase::initialSetup` because of the early return in `Steady::init`

**Comment by YaqiWang** [comment]
> we did not hit this before because we never proceeded to read the restartable data in `FEProblemBase::initialSetup` because of the early return in `Steady::init`

Yes.

**Comment by lindsayad** [comment]
I'm coming around to your changes. In the vein of supporting more things, I would vote to perform the refinements on the subapp upon recover instead of erroring. I mean, why not?

File: modules/fluid_properties/test/tests/two_phase_fluid_properties_independent/tests
Comment (lindsayad) [reply modules/fluid_properties/test/tests/two_phase_fluid_properties_independent/tests:26]: Ah I see that because this uses the same input as the one above that it messed up the checkpoint files (which I think you might have said somewhere else)

**Comment by YaqiWang** [comment]
One argument of having that error is that we have another error above it for restart, which I do not quite understand.

File: modules/fluid_properties/test/tests/two_phase_fluid_properties_independent/tests
Comment (YaqiWang) [reply modules/fluid_properties/test/tests/two_phase_fluid_properties_independent/tests:26]: Ah, yes, forgot to mention that. I added a comment above this line in the tests file ;-)

**Comment by lindsayad** [comment]
I don't think that not understanding why one feature is disallowed is an argument to disallow a feature we do understand. We may not write out the mesh for subapps but we do write out the subapp solution vectors into our recoverable/restartable data and because we are doing recovery (as opposed to restart) we can expect that all the input files we are running with are the same as before we recovered. Hence we can trust that a refined subapp mesh (post recover) will match our recovered subapp solution vectors

**Comment by YaqiWang** [comment]
Will try it a little later.

**Comment by lindsayad** [comment]
I checked that with your patch all MOOSE test pass locally with `--recover` and with `--recover -p2 --distributed-mesh`

**Comment by YaqiWang** [comment]
@lindsayad Add a comment to do uniform refinement for subapps when recovering. I hope we can check applications before we merge this. We might need to fix few things like what we did here for framework and modules. Looks like currently only Bison, R7, and Griffin has recovery tests. I confirmed that Griffin works fine with this change.

File: framework/src/problems/FEProblemBase.C
Comment (lindsayad) [inline framework/src/problems/FEProblemBase.C:756]: If you are going to add this error, you should comment on why this is not supported. The above error is somewhat of a mystery; we do not want to add another mystery

**Review by lindsayad — DISMISSED** [review]

File: framework/src/problems/FEProblemBase.C
Comment (lindsayad) [reply framework/src/problems/FEProblemBase.C:756]: I believe the restart error was added in https://github.com/idaholab/moose/pull/5693 and @friedmud created an issue for it in https://github.com/idaholab/moose/issues/5691

Comment (YaqiWang) [reply framework/src/problems/FEProblemBase.C]: I think we must include this error. I added an inline comment to make clear what it is for.

Comment (YaqiWang) [reply framework/src/problems/FEProblemBase.C]: If we add the flexible restart capability for addressing #21011, I think we can remove the Exodus restart completely and this error.

Comment (lindsayad) [reply framework/src/problems/FEProblemBase.C]: exodus file restart will never happen when recovering:
```c++
  if ((_app.isRestarting() || _app.isRecovering()) && (_app.isUltimateMaster() || _force_restart))
  {
    TIME_SECTION("restartFromFile", 3, "Restarting From File");

    _restart_io->readRestartableDataHeader(true);
    _restart_io->restartEquationSystemsObject();

    /**                                                                                                               
     * TODO: Move the RestartableDataIO call to reload data here. Only a few tests fail when doing                    
     * this now. Material Properties aren't sized properly at this point and fail across the board,                   
     * there are a few other misc tests that fail too.                                                                
     *                                                                                                                
     * _restart_io->readRestartableData();                                                                            
     */
  }
  else
  {
    ExodusII_IO * reader = _app.getExReaderForRestart();

    if (reader)
    {
      TIME_SECTION("copyingFromExodus", 3, "Copying Variables From Exodus");

      _nl->copyVars(*reader);
      _aux->copyVars(*reader);
    }
```

Comment (lindsayad) [reply framework/src/problems/FEProblemBase.C]: My understanding of the code is this:

When either `_app.isRecovering()` or `_app.isRestarting()`, we are just going to write over the subapp solution vector when we restore the cached subapp backup; the restoration of the backup occurs after where we copy variables from exodus so it doesn't matter what the solution held by the exodus file was. What I think we should do instead is simply not try to read exodus solutions when we are a subapp and `_app.isRecovering() || _app.isRestarting()`

Comment (YaqiWang) [reply framework/src/problems/FEProblemBase.C]: Do you have specific code changes for this? I think I agree with you.

Comment (YaqiWang) [reply framework/src/problems/FEProblemBase.C]: If we can improve `libMesh::ExodusII_IO` two functions `copy_nodal_solution` and `copy_elemental_solution` to support a uniform refinement level argument, we will not have this mess.

File: framework/src/problems/FEProblemBase.C
Comment (YaqiWang) [inline framework/src/problems/FEProblemBase.C:682]: Double `((` parenthesis.

File: framework/src/problems/FEProblemBase.C
Comment (lindsayad) [reply framework/src/problems/FEProblemBase.C:684]: It makes it so that if you are recovering or restarting you don't proceed to the `else` branch with exodus restart

Comment (YaqiWang) [reply framework/src/problems/FEProblemBase.C:684]: Right, I missed that.

**Comment by lindsayad** [comment]
>  I confirmed that Griffin works fine with this change.

I just ran Griffin recover tests and found two failures
```
stochastic.CNM/cli_normal .......................................................... [recover] FAILED (ERRMSG)
stochastic.CNM/transfer_normal ..................................................... [recover] FAILED (ERRMSG)
```
both seemingly due to perf graph storing/loading
```
stochastic.CNM/cli_normal: *** ERROR ***
stochastic.CNM/cli_normal: Section Name Not Found: 
stochastic.CNM/cli_normal: 
stochastic.CNM/cli_normal: Stack frames: 16
stochastic.CNM/cli_normal: 0: libMesh::print_trace(std::ostream&)
stochastic.CNM/cli_normal: 1: moose::internal::mooseErrorRaw(std::__cxx11::basic_string<char, std::char_traits<char>, std::allocator<char> >, std::__cxx11::basic_string<char, std::char_traits<char>, std::allocator<char> >)
stochastic.CNM/cli_normal: 2: void mooseError<char const (&) [25], std::__cxx11::basic_string<char, std::char_traits<char>, std::allocator<char> > const&>(char const (&) [25], std::__cxx11::basic_string<char, std::char_traits<char>, std::allocator<char> > const&)
stochastic.CNM/cli_normal: 3: /home/lindad/projects/moose/framework/libmoose-devel.so.0(+0xa7fb2f) [0x7f1b3607fb2f]
stochastic.CNM/cli_normal: 4: dataLoad(std::istream&, std::unique_ptr<PerfNode, std::default_delete<PerfNode> > const&, void*)
stochastic.CNM/cli_normal: 5: dataLoad(std::istream&, PerfGraph&, void*)
stochastic.CNM/cli_normal: 6: RestartableDataIO::deserializeRestartableData(std::unordered_map<std::__cxx11::basic_string<char, std::char_traits<char>, std::allocator<char> >, RestartableDataValuePair, std::hash<std::__cxx11::basic_string<char, std::char_traits<char>, std::allocator<char> > >, std::equal_to<std::__cxx11::basic_string<char, std::char_traits<char>, std::allocator<char> > >, std::allocator<std::pair<std::__cxx11::basic_string<char, std::char_traits<char>, std::allocator<char> > const, RestartableDataValuePair> > > const&, std::istream&, std::unordered_set<std::__cxx11::basic_string<char, std::char_traits<char>, std::allocator<char> >, std::hash<std::__cxx11::basic_string<char, std::char_traits<char>, std::allocator<char> > >, std::equal_to<std::__cxx11::basic_string<char, std::char_traits<char>, std::allocator<char> > >, std::allocator<std::__cxx11::basic_string<char, std::char_traits<char>, std::allocator<char> > > > const&)
stochastic.CNM/cli_normal: 7: RestartableDataIO::readRestartableData(std::unordered_map<std::__cxx11::basic_string<char, std::char_traits<char>, std::allocator<char> >, RestartableDataValuePair, std::hash<std::__cxx11::basic_string<char, std::char_traits<char>, std::allocator<char> > >, std::equal_to<std::__cxx11::basic_string<char, std::char_traits<char>, std::allocator<char> > >, std::allocator<std::pair<std::__cxx11::basic_string<char, std::char_traits<char>, std::allocator<char> > const, RestartableDataValuePair> > > const&, std::unordered_set<std::__cxx11::basic_string<char, std::char_traits<char>, std::allocator<char> >, std::hash<std::__cxx11::basic_string<char, std::char_traits<char>, std::allocator<char> > >, std::equal_to<std::__cxx11::basic_string<char, std::char_traits<char>, std::allocator<char> > >, std::allocator<std::__cxx11::basic_string<char, std::char_traits<char>, std::allocator<char> > > > const&, unsigned int)
stochastic.CNM/cli_normal: 8: RestartableDataIO::readRestartableData(std::vector<std::unordered_map<std::__cxx11::basic_string<char, std::char_traits<char>, std::allocator<char> >, RestartableDataValuePair, std::hash<std::__cxx11::basic_string<char, std::char_traits<char>, std::allocator<char> > >, std::equal_to<std::__cxx11::basic_string<char, std::char_traits<char>, std::allocator<char> > >, std::allocator<std::pair<std::__cxx11::basic_string<char, std::char_traits<char>, std::allocator<char> > const, RestartableDataValuePair> > >, std::allocator<std::unordered_map<std::__cxx11::basic_string<char, std::char_traits<char>, std::allocator<char> >, RestartableDataValuePair, std::hash<std::__cxx11::basic_string<char, std::char_traits<char>, std::allocator<char> > >, std::equal_to<std::__cxx11::basic_string<char, std::char_traits<char>, std::allocator<char> > >, std::allocator<std::pair<std::__cxx11::basic_string<char, std::char_traits<char>, std::allocator<char> > const, RestartableDataValuePair> > > > > const&, std::unordered_set<std::__cxx11::basic_string<char, std::char_traits<char>, std::allocator<char> >, std::hash<std::__cxx11::basic_string<char, std::char_traits<char>, std::allocator<char> > >, std::equal_to<std::__cxx11::basic_string<char, std::char_traits<char>, std::allocator<char> > >, std::allocator<std::__cxx11::basic_string<char, std::char_traits<char>, std::allocator<char> > > > const&)
stochastic.CNM/cli_normal: 9: FEProblemBase::initialSetup()
stochastic.CNM/cli_normal: 10: MooseApp::executeExecutioner()
stochastic.CNM/cli_normal: 11: MooseApp::run()
stochastic.CNM/cli_normal: 12: main
stochastic.CNM/cli_normal: 13: /lib/x86_64-linux-gnu/libc.so.6(+0x29d90) [0x7f1b2d829d90]
stochastic.CNM/cli_normal: 14: __libc_start_main
stochastic.CNM/cli_normal: 15: /home/lindad/projects/griffin/griffin-devel(+0x3755) [0x5614a7215755]
stochastic.CNM/cli_normal: application called MPI_Abort(MPI_COMM_WORLD, 1) - process 0
stochastic.CNM/cli_normal: [unset]: write_line error; fd=-1 buf=:cmd=abort exitcode=1
```
Looking at this in the debugger it looks like there is a nonsensical value for `num_children` in 
```c++
void
dataLoad(std::istream & stream, const std::unique_ptr<PerfNode> & node, void * perf_graph)
```

**Comment by YaqiWang** [comment]
Let me try. I did not see this though.

**Comment by lindsayad** [comment]
Silly me. Just properly setting the `prereq` for those tests fixed the failures

**Comment by YaqiWang** [comment]
I really hope we can improve `SystemBase::copyVars` so that we do not have to do late uniform refinement but that seems quite difficult. But since we may abandon Exodus restart, I guess we do not have to do it.

**Comment by lindsayad** [comment]
requesting alternative reviewers since I'm not a partial author on this PR

**Comment by YaqiWang** [comment]
> 

I thought tests in one folder are always run sequentially. Is there are setting for not doing that?

File: modules/fluid_properties/test/tests/two_phase_fluid_properties_independent/tests
Comment (GiudGiud) [reply modules/fluid_properties/test/tests/two_phase_fluid_properties_independent/tests:26]: cant you use prereq to stagger them? Or is this not obeyed for recover?

**Review by GiudGiud — CHANGES_REQUESTED** [review]
declareFeatureSupported("recover", ...) 
would be so nice and cleaner way to handle this

File: test/tests/restart/pointer_restart_errors/tests
Comment (YaqiWang) [reply test/tests/restart/pointer_restart_errors/tests:20]: This test is only used for generating the checkpoint file. If we do not add `recover = false` for it, it runs in recover mode and now we will hit an error for a custom type that is missing `DataLoad` on purpose because we do not have early return in `init` as before.

File: modules/fluid_properties/test/tests/two_phase_fluid_properties_independent/tests
Comment (YaqiWang) [reply modules/fluid_properties/test/tests/two_phase_fluid_properties_independent/tests:26]: Possibly but I would say that this test is possible not intended to run in recover and skipping it is fun because it probably does not test anything new.

File: test/tests/restart/pointer_restart_errors/tests
Comment (GiudGiud) [reply test/tests/restart/pointer_restart_errors/tests:20]: ok works for me

**Comment by GiudGiud** [comment]
Tests in a folder are not run sequentially I think.
Otherwise we would not need prereq.
and also the timings would be currently wrong
For example:
```
Ran 133 tests in 4.7 seconds. Average test time 0.2 seconds, maximum test time 0.3 seconds.
```

File: modules/fluid_properties/test/tests/two_phase_fluid_properties_independent/tests
Comment (GiudGiud) [inline modules/fluid_properties/test/tests/two_phase_fluid_properties_independent/tests:26]:
```suggestion
    prereq = test
```

**Comment by GiudGiud** [comment]
Trying something, see if it passes

**Comment by GiudGiud** [comment]
Test failure is unrelated.
This is good to go once my proposed changes are squashed into an earlier commit

**Comment by lindsayad** [comment]
> I thought tests in one folder are always run sequentially. Is there a setting for not doing that?

They are on CIVET. But with CIVET without `prereq`, you run all the part 1s first, and then run all the part 2s. This is problematic if multiple tests potentially write to the same restart/recover files. When you specify `prereq` CIVET is smart and will run part1 and part2 of the prereq before moving to the dependent test

**Comment by YaqiWang** [comment]
@friedmud

File: framework/src/problems/FEProblemBase.C
Comment (lindsayad) [reply framework/src/problems/FEProblemBase.C:600]:
```suggestion
      "Checkpoint recovery and restart and exodus restart are all mutually exclusive.");
```

File: modules/fluid_properties/test/tests/two_phase_fluid_properties_independent/tests
Comment (lindsayad) [reply modules/fluid_properties/test/tests/two_phase_fluid_properties_independent/tests]: Looks like we edited this to add `prereq` 👍

**Review by GiudGiud — APPROVED** [review]

**Pull request merged**
